tt_metal/impl/buffers/buffer.cpp

```diff
@@ -198,89 +227,158 @@ BufferPageMapping generate_buffer_page_mapping(const Buffer &buffer) {
-Buffer::Buffer(Buffer &&other) :
-    device_(other.device_),
-    size_(other.size_),
-    address_(other.address_),
-    page_size_(other.page_size_),
-    buffer_type_(other.buffer_type_),
-    buffer_layout_(other.buffer_layout_),
-    shard_parameters_(std::move(other.shard_parameters_)),
-    bottom_up_(other.bottom_up_),
-    buffer_page_mapping_(std::move(other.buffer_page_mapping_)),
-    allocate_(other.allocate_) {
-    // Set `other.device_` to be nullptr so destroying other does not deallocate reserved address space that is
-    // transferred to `this`
-    other.device_ = nullptr;
-}
-
-Buffer &Buffer::operator=(Buffer &&other) {
-    if (this != &other) {
-        this->device_ = other.device_;
-        this->size_ = other.size_;
-        this->address_ = other.address_;
-        this->page_size_ = other.page_size_;
-        this->buffer_type_ = other.buffer_type_;
-        this->buffer_layout_ = other.buffer_layout_;
-        this->shard_parameters_ = std::move(other.shard_parameters_);
-        this->bottom_up_ = other.bottom_up_;
-        this->buffer_page_mapping_ = std::move(other.buffer_page_mapping_);
-        this->allocate_ = other.allocate_;
-        // Set `other.device_` to be nullptr so destroying other does not deallocate reserved address space that is
-        // transferred to `this`
-        other.device_ = nullptr;
+
+    device_->push_work([self = weak_self.lock()] {
+        // Because the status is DEALLOCATION_REQUESTED, it won't be changed by anyone else, no need to lock a mutex
+        if (!self->device_->initialized_) {
+            return;
+        }
+
+        detail::BUFFER_MAP.erase({self->device()->id(), self->address()});
+        detail::DeallocateBuffer(self.get());
+        self->allocation_status_.store(AllocationStatus::DEALLOCATED, std::memory_order::relaxed);
+    });
+}
+
+void Buffer::deallocateAndDelete(Buffer* buffer) {
+    // This is the last reference to the buffer, no need to lock or update AllocationStatus
+    buffer->device_->push_work([buffer] {
+        // Buffer will be deleted at the end of this block
+        std::unique_ptr<Buffer> unique_buffer = std::unique_ptr<Buffer>(buffer);
+
+        auto status = buffer->allocation_status_.load(std::memory_order::relaxed);
+        if (status == AllocationStatus::NOT_ALLOCATED || status == AllocationStatus::ALLOCATION_REQUESTED || status == AllocationStatus::DEALLOCATED) {
+            // Buffer isn't allocated, nothing to be done
+            return;
+        }
+
+        if (!buffer->device_->initialized_ || buffer->size_ == 0) {
+            return;
+        }
+
+        detail::BUFFER_MAP.erase({buffer->device_->id(), buffer->address_});
+        detail::DeallocateBuffer(buffer);
+    });
+}
+
+bool Buffer::is_allocated() const {
+    auto allocation_status = allocation_status_.load(std::memory_order::relaxed);
+    if (device_->can_use_passthrough_scheduling()) {
+        return allocation_status == AllocationStatus::ALLOCATED;
```

Comment:
Is this branch needed? If we're running single threaded, each time this fn is called the buffer will either be allocated or deallocated. I don't think the requested state will be hit. I'm concerned about host perf with branching.

Reply:
You may loose near 10-20 clock cycles with branching (3-4 ns). You should not worry about it in places like this.

Reply:
If there's no reason logical to have it, lets get rid of it please. Thanks!

Reply:
@tt-asaigal Having this branch is important, otherwise the behavior of this method would be incorrect.

Imagine that you're allocating a buffer from another thread, resulting in scheduling the allocation and the state being ALLOCATION_REQUESTED. Some other operation running in the worker thread can call `buffer->is_allocated()` before the allocation was executed.

If you're running single-threaded you wouldn't see a performance hit from this check, because in this case branch predictor will have a roughly 100% hit on predicting this check.